engines/dotc/src/main/scala/scala/macros/internal/engines/dotc/DottyUniverse.scala

```diff
@@ -153,98 +135,147 @@ case class DottyUniverse(prefix: untpd.Tree)(implicit ctx: Context) extends macr
+      decltpe: Option[TypeTree],
       body: Term
-  ): Defn = {
+  ): Defn =
     untpd.DefDef(
-      name.asInstanceOf[untpd.Ident].name.asTermName,
+      name.toTermName,
       tparams,
       paramss,
       decltpe.getOrElse(untpd.TypeTree()),
       body
     )
-  }
 
   def DefnObject(
       mods: List[Mod],
-      name: TermName,
+      name: String,
       templ: Template
   ): Defn =
     untpd
       .ModuleDef(
-        name.asInstanceOf[untpd.Ident].name.asTermName,
+        name.toTermName,
         templ
       )
       .autoPos
 
-  def PatVar(name: TermName): PatVar = name
-
   // =========
+  // typed trees
+  // =========
+  object typed extends typedApi {
+    type Tree = tpd.Tree
+    type Term = tpd.Tree
+    type Def = tpd.Tree
+
+    def treePosition(tree: Tree): Position = Position(tree.pos)
+    def treeSyntax(tree: Tree): String = tree.show
+    def treeStructure(tree: Tree): String = unsupported
+
+    def symOf(tree: Def): Symbol = tree.symbol
+    def typeOf(tree: Term): Type = tree.tpe
+    def ref(sym: Symbol): Term = tpd.ref(sym)
+
+    // only for terms, no extractor for type trees
+    def NameUnapply(tree: Tree): Option[Denotation] = tree match {
+      case id: tpd.Ident if id.name.isTermName => Some(id.tpe.asInstanceOf[Types.NamedType].denot)
+      case _ => None
+    }
+
+    def Select(qual: Term, name: String): Term = tpd.Select(qual, name.toTermName)
+    def SelectUnapply(tree: Tree): Option[(Term, Symbol)] = tree match {
+      case tpd.Select(qual, name) if name.isTermName => Some((qual, tree.tpe.termSymbol))
+      case _ => None
+    }
+
+    def Apply(fun: Term, args: List[Term]): Term = tpd.Apply(fun, args)
+    def ApplyUnapply(tree: Tree): Option[(Term, List[Term])] = tree match {
+      case tpd.Apply(fun, args) => Some((fun, args))
+      case _ => None
+    }
+
+    def FunctionUnapply(tree: Tree): Option[(List[Symbol], Term)] = tree match {
+      case tpd.Block(Nil, body) => FunctionUnapply(body)
+      case tpd.Block((meth: tpd.DefDef) :: Nil, _: tpd.Closure) if meth.name == nme.ANON_FUN =>
+        Some((meth.vparamss.head.map(_.symbol), meth.rhs))
+      case _ => None
+    }
+
+    def If(cond: Term, thenp: Term, elsep: Term): Term = tpd.If(cond, thenp, elsep)
+    def ValDef(rhs: Term, tpOpt: Option[Type], mutable: Boolean): Tree = {
+      val flags = if (mutable) Flags.Mutable else Flags.EmptyFlags
+      val vsym = ctx.newSymbol(
+        ctx.owner,
+        NameKinds.UniqueName.fresh("temp".toTermName),
+        flags,
+        tpOpt.getOrElse(rhs.tpe)
+      )
+      tpd.ValDef(vsym, rhs)
+    }
+
+    def transform(tree: Tree)(pf: PartialFunction[Tree, Tree]): Tree = {
+      new tpd.TreeMap() {
+        override def transform(tree: tpd.Tree)(implicit ctx: Context) = {
+          def updateOwner(subtree: tpd.Tree) = ensureOwner(subtree, ctx.owner)
+          pf.lift(tree).map(updateOwner(_)).getOrElse(super.transform(tree))
+        }
+      }.transform(tree)
+    }
```

Comment:
As a note, I find this API not convenient for usage if I want to recur in a pattern case. Maybe the compiler-side API should be more general. The user-side API can keep as it's for now.

Reply:
I removed transform from the API since it's not used. We can take inspiration from scala-reflect TypingTransformer that gives fine-grained control over recursion.